### PR TITLE
Sonarcloud static analysis

### PR DESCRIPTION
**Pull request by RogerZhongAWS**
### Motivation
- Enables sonarqube to analyze our code for bugs, security, and style issues
#### Change summary
- Fixes for most of the issues identified by sonarqube
- Autoformatting changes

#### Revision diff summary
If there is more than one revision, please explain what has been changed since the last revision.

### Testing
 **Is your change tested? If not, please justify the reason.**  
 **Please list your testing steps and test results.** 
- CI test run result: <link>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by BolongZhang-AWS — DISMISSED** [review]
Can we add some documentation for how to navigate and use Sonarcloud?

File: .github/workflows/sonarcloud.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    branches:
+      - '*'
+  pull_request:
+    branches:
+      - '*'
+
+jobs:
+  sonarcloud:
+    runs-on: ubuntu-latest
+    if: (github.event_name == 'push') || ((github.event_name == 'pull_request') && (github.event.pull_request.head.repo.full_name != github.repository))
```
Comment (shangabl) [inline .github/workflows/sonarcloud.yml:13]: The first two conditionals are redundant I think?

**Review by shangabl — COMMENTED** [review]
Can this be run locally? It would be good to have a script like we do for checking format and cppcheck so when we are developing we can just mount to a docker container and run the scripts.

File: .github/workflows/sonarcloud.yml (hunk 1 above)
Comment (RogerZhongAWS) [reply .github/workflows/sonarcloud.yml:13]: they are needed to prevent duplicate runs from triggering if a new update is made to a PR

**Comment by RogerZhongAWS** [comment]
> Can this be run locally? It would be good to have a script like we do for checking format and cppcheck so when we are developing we can just mount to a docker container and run the scripts.

yes, its pretty straightforward.

This PR pretty much already contains the script for doing so. You need to:
1. Add the compile flag that disables optimization "-O0" and enable -fprofile-arcs -ftest-coverage which generates control flow information
2. Run the tests as you normally would.
3. Find the directory where the .gcda files are stored. This is by default the same directory that the object files *.o are written to
4. Run the gcov command with a file name or for a directory, use the same command in the sonarcloud.yml file to convert the .gcda into a human readable output.
5. Add whatever other steps you want to collate the data.

**Comment by shangabl** [comment]
> > Can this be run locally? It would be good to have a script like we do for checking format and cppcheck so when we are developing we can just mount to a docker container and run the scripts.
> 
> yes, its pretty straightforward.
> 
> This PR pretty much already contains the script for doing so. You need to:
> 
> 1. Add the compile flag that disables optimization "-O0" and enable -fprofile-arcs -ftest-coverage which generates control flow information in the form of a .gcno file
> 2. Run the tests as you normally would. This will write encoded coverage information to .gcda files
> 3. Find the directory where the .gcda files are stored. This is by default the same directory that the object files *.o are written to
> 4. Run the gcov command with a file name or for a directory, use the same command in the sonarcloud.yml file to convert the .gcda into a human readable output.
> 5. Add whatever other steps you want to collate the data.

I think it would be good to have a bash script for this

**Review by shangabl — APPROVED** [review]

**Review by xlcheng1 — APPROVED** [review]

**Pull request merged**